docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,548 @@
+2. Write Only -> Done: Update the new created table state to `public`.
+
+In step-1, we need update some table info in one schema-version. Technically, we can implement is since we already support `ActionCreateTables` DDL job.
+
+### Alter Table Add Foreign Key
+
+Here is an example:
+
+```sql
+create table t1 (id int key,a int, index(a));
+create table t2 (id int key,a int);
+alter  table t2 add foreign key fk(a) references t1(id) ON DELETE CASCADE;
+```
+
+Just like create table, we should validate first, and return error if the conditions for creating foreign keys are not met, and also need to do double-check.
+
+When build `TableInfo`, we need to auto create an index for foreign key columns if there is no index cover foreign key columns.
+And this is divides the problem into two cases:
+- Case-1: No need to auto create index, and only add foreign key constrain.
+- Case-2: Need auto create index for foreign key
+
+#### Case-1: Only add foreign key constrain
+
+The DDL owner handle add foreign key constrain step is:
+
+1. None -> Write Only: add foreign key constrain which state is `write-only` into table and update the reference table info.
+3. Write Only - Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:
+   ```sql
+   select count(*) from t2 where t2.a not in (select id from t1);
+   ```
+   The expected result is `0`, otherwise, an error is returned and cancel the ddl job.
+4. Write Reorg -> Public: update the foreign key constrain state to `public`.
+
+DML should also check the foreign key constrain which state is `write-only`
+
+#### Case-2: Auto create index for foreign key and add foreign key constrain
+
+As TiDB support multi-schema change now, we can split this into 2 sub-ddl job.
+- Add Index DDL job
+- Add Foreign Key Constrain DDL job
+
+We should do add index DDL job first, after index ddl job finish `write-reorg` and ready for public, then start to do add foreign key constrain ddl job.
+
+### Drop Table
+
+If `foreign_key_checks` is `ON`, then drop the table which has foreign key references will be rejected.
+
+```sql
+> drop table t1;
+(3730, "Cannot drop table 't1' referenced by a foreign key constraint 't2_ibfk_1' on table 't2'.")
+```
+
+### Drop Index
+
+Drop index which used by foreign key will be rejected.
+
+```sql
+> set @@foreign_key_checks=0; -- Even disable foreign_key_checks, you still can't drop the index which used for foreign key constrain.
+Query OK, 0 rows affected
+> alter table t2 drop index fk;
+(1553, "Cannot drop index 'fk': needed in a foreign key constraint")
+```
+
+### Rename Table/Column
+
+Rename table which has foreign key references, should also need to update the related child table info.
+
+```sql
+create table t1 (id int key,a int, index(a));
+create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);
+rename table t1 to t11;
```

Comment:
why not just let `FKInfo` only store the refered table id

Reply:
truncate table / modify column with reorg  -> change child table id.

Reply:
After trying, you are right, I think recorded table id and schema id are better here.

Reply:
em... We still need to record the table/schema name, not table/schema id. Here is an example:

```sql
mysql>set @@foreign_key_checks=0;
Query OK, 0 rows affected
Time: 0.000s
mysql>create table t2 (a int key, foreign key fk(a) references t1(id));
Query OK, 0 rows affected
Time: 0.011s
mysql>create table t1 (id int key);
Query OK, 0 rows affected
Time: 0.011s
mysql>set @@foreign_key_checks=1;
Query OK, 0 rows affected
Time: 0.000s
mysql>insert into t2 values (1);
(1452, 'Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t1` (`id`))')
```

As you can see, the table `t2` is refer to table `t1`, and when create table `t2`, the table `t1` still not be created, so we can't know table `t1`.